org.metaborg.spoofax.core/src/main/java/org/metaborg/spoofax/core/completion/JSGLRCompletionService.java

```diff
@@ -2,84 +2,752 @@
+                        // expected two lists
+                        if(change.getSubtermCount() != 2 || !(oldNode instanceof IStrategoList)
+                            || !(newNode instanceof IStrategoList)) {
+                            logger.error("Unexpected proposal term {}, skipping", proposalTerm);
+                            continue;
+                        }
+
+                        ITokenizer tokenizer = ImploderAttachment.getTokenizer(oldNode);
+                        final ImploderAttachment oldListIA = oldNode.getAttachment(ImploderAttachment.TYPE);
+                        insertionPoint = position;
+
+                        // if list is empty
+                        // insert after the first non-layout token before the leftmost token of the completion
+                        // node
+                        if(((IStrategoList) oldNode).getSubtermCount() == 0) {
+                            int tokenPosition =
+                                oldListIA.getLeftToken().getIndex() - 1 > 0 ? oldListIA.getLeftToken().getIndex() - 1
+                                    : 0;
+                            while((tokenizer.getTokenAt(tokenPosition).getKind() == IToken.TK_LAYOUT || tokenizer
+                                .getTokenAt(tokenPosition).getKind() == IToken.TK_ERROR) && tokenPosition > 0)
+                                tokenPosition--;
+                            insertionPoint = tokenizer.getTokenAt(tokenPosition).getEndOffset();
+                        } else {
+                            // if list is not empty
+                            // insert after at end offset of the rightmost token of the element before the
+                            // completion
+                            StrategoTerm elementBefore =
+                                (StrategoTerm) oldNode.getSubterm(oldNode.getAllSubterms().length - 1);
+                            insertionPoint =
+                                elementBefore.getAttachment(ImploderAttachment.TYPE).getRightToken().getEndOffset();
+                        }
+                        suffixPoint = insertionPoint + 1;
+
+                    } else if(change.getConstructor().getName().contains("INSERT_BEFORE")) {
+                        // expect two terms and 1st should be an element of a list
+                        final StrategoTerm oldList = (StrategoTerm) ParentAttachment.getParent(oldNode);
+                        ITokenizer tokenizer = ImploderAttachment.getTokenizer(oldNode);
+                        if(change.getSubtermCount() != 2 || !(oldNode instanceof IStrategoAppl)
+                            || !(newNode instanceof IStrategoList) || !(oldList instanceof IStrategoList)) {
+                            logger.error("Unexpected proposal term {}, skipping", proposalTerm);
+                            continue;
+                        }
+
+
+                        IStrategoTerm[] subterms = ((IStrategoList) oldList).getAllSubterms();
```

Comment:
Calling `getAllSubterms` on a list copies the terms into an array, which could be expensive.